Model+dto+context

KisMiska opened the pull request:
Layers are connected so a request goes al the way to cosmosdDb (from Postman, not frontend). Endpoints working for creating the questionnaires based on the given metadata, and also the deletion.

valid body for endpoint .../api/surveys (POST)
[metadata.json](https://github.com/user-attachments/files/21933867/metadata.json)

KisMiska committed:
model

KisMiska committed:
...

KisMiska committed:
dtos created and flexibility added to the model

KisMiska committed:
...

KisMiska committed:
unneccessary lines deleted

KisMiska committed:
cherry

KisMiska committed:
connectionString in env and set up

KisMiska committed:
azure,service,repo pipeline and preapred for extesion classes

KisMiska committed:
fixup! connectionString in env and set up

SzLRichard committed:
Added ToDto and ToModel mappers

KisMiska committed:
service and repo put in together

SzLRichard committed:
Merge pull request #71 from Explorer-Consulting/68-add-dto-modell-mapping

Added ToDto and ToModel mappers

KisMiska committed:
layers connected

SzLRichard committed:
Implemented delete endpoint

SzLRichard committed:
Added implementation

KisMiska committed:
context changes

KisMiska committed:
models adjuted to newtonsoft

SzLRichard committed:
Deleting survey metadata works

SzLRichard committed:
Minor changes

KisMiska committed:
surveycompiler done

SzLRichard committed:
Merge branch 'QuestionnaireDeletion' into model+dto+context

KisMiska committed:
... nothing extra

KisMiska committed:
...

SzLRichard committed:
Improved creation exception handling

SzLRichard committed:
Removed cancellation tokens

SzLRichard committed:
Configured update dtos

KisMiska committed:
Merge branch 'model+dto+context' of https://github.com/Explorer-Consulting/SchoolFeedBackWebAPI into model+dto+context

# Conflicts:
#	FeedBackApp.Backend/Application/DTOs/QuestionnaireDTOs/CreationResponseDTO.cs

KisMiska committed:
exception handling

SzLRichard committed:
Gave correct types to dates

SzLRichard committed:
Added basic survey data validation

SzLRichard committed:
Wired the validators

SzLRichard committed:
Improved validation error messages

KisMiska committed:
model changed, question are in separated co for easier access

SzLRichard committed:
Merge branch 'CreateValidation' into model+dto+context

SzLRichard committed:
Modified validator to match new model

KisMiska committed:
delete updated

KisMiska committed:
Merge branch 'model+dto+context' of https://github.com/Explorer-Consulting/SchoolFeedBackWebAPI into model+dto+context

KisMiska committed:
Merge branch 'develop' into model+dto+context

BHunor-ex commented on FeedBackApp.Backend/Application/DTOs/QuestionnaireDTOs/CreateSurveyMetadataDto.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using Newtonsoft.Json;
+
+namespace Application.DTOs.QuestionnaireDTOs
```
nit: This **DTOs.QuestionareDtos** is a bit redundant, **DTOs.Questionnare** would be enough.

Also in the file names I can see `DTO` and `Dto` endings as well, choose one and stick to it.

BHunor-ex commented on FeedBackApp.Backend/Application/DTOs/QuestionnaireDTOs/DeletionResponseDTO.cs:
```diff
@@ -0,0 +1,15 @@
+    public class DeletionResponseDTO
+    {
+        public bool Success { get; set; }
+        public string Message { get; set; }=string.Empty;
```
format this line `... { get; set; } = string.Empty;`

BHunor-ex commented on FeedBackApp.Backend/Application/DTOs/QuestionnaireDTOs/QuestionTemplateDto.cs:
```diff
@@ -0,0 +1,19 @@
+
+        [JsonProperty("type")]
+        [JsonConverter(typeof(StringEnumConverter))]
+        public QuestionType? Type { get; set; }
```
Why is this `QuestionType?` (why can it be nullable?) can a question have no Type? Anyway if there is a possibility that a question has no type, it is better to set a `QuestionType.NoType =  0` in the Enum, so it becomes clearer.

SzLRichard replied:
The reason I allowed null value for the type, is because when we deserialize the JSON into the DTO, if the QuestionType was not provided, it would automatically set it to the value 0, which is a valid QuestionType in the enum.
However, if I define it as "QuestionType?", then it defaults to null, and this allows me to throw an error when validating the dto's fields. @BHunor-ex

BHunor-ex replied:
Then why is the QuestionType is not provided from the frontend? I mean this is why we don't allow null values, and set the QuestonType.Unknown or .NoType to 0, this way it will automatically set it to 0 (or Unkown) and you can detect that in the validator (or anywhere else you need), and it's much more straightforward what is happening.

SzLRichard replied:
The frontend ideally always will provide the QuestionType, but if something goes wrong, I want to be able to detect the problem. I will add the Unknown type, that sounds good.

BHunor-ex commented on FeedBackApp.Backend/Application/DTOs/QuestionnaireDTOs/QuestionTemplateDto.cs:
```diff
@@ -0,0 +1,19 @@
+        public QuestionType? Type { get; set; }
+
+        [JsonProperty("answerOptions")]
+        public IList<string>? AnswerOptions { get; set; } = new List<string>();
```
Why can it be nullable? It is possible that an answer has no options? Anyway if it has no options, you can still make it as as an new `EmptyList` as you do.

KisMiska replied:
There are open ended questions, in which the student needs to write a comment/review, so there are no options to choose from. So i thought i would not store a empty key value pair in the Db

BHunor-ex replied:
Maybe there is something i don't understand right now.

Anyway we shouldn't send null values to the backend, only if they are **Optional** values (e.g. we wan't to allow the frontend to call our endpoints with variable data). I this case I think, we could still send an EmptyList with the dto (this doesn't mean we need to store it in the DB.)

BHunor-ex commented on FeedBackApp.Backend/Application/Extensions/QuestionnaireExtensions/QuestionnaireMapping.cs:
```diff
@@ -0,0 +1,138 @@
+
+namespace Application.Extensions.QuestionnaireExtensions
+{
+    public static class SurveyMetadataMapper
```
Maybe **SurveyMetadataMappingExtensions** would be a more appropriate name.

BHunor-ex commented on FeedBackApp.Backend/Application/Services/QuestionnaireService.cs:
```diff
@@ -1,64 +1,75 @@
     {
         private readonly IQuestionnaireRepository _repository;
-        public QuestionnaireService(IQuestionnaireRepository repository)
+        IValidator<CreateSurveyMetadataDto> _validator;
```
this should be private i think.

BHunor-ex commented on FeedBackApp.Backend/Application/Services/QuestionnaireService.cs:
```diff
@@ -1,64 +1,75 @@
 
-                        await _repository.SaveQuestionnaireAsync(questionnaire);
-                    }
+                if (surveyDeleted || questionnairesDeleted || questionTemplateDeleted)
```
Are we sure if any of these tree is deleted correctly and the others are not (so it could be `true || false || false`) we want to send a successful message?

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs:
```diff
@@ -31,22 +33,33 @@ public sealed class QuestionnaireCompilerWorkerEncapsulator(IQuestionnaireServic
+
+                if (dto == null)
+                {
+                    var badResponse = request.CreateResponse(HttpStatusCode.BadRequest);
```
Should use the _logger to prompt a message to the console, so if anything happens you can detect it immediately.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs:
```diff
@@ -31,22 +33,33 @@ public sealed class QuestionnaireCompilerWorkerEncapsulator(IQuestionnaireServic
             {
-                response.StatusCode = HttpStatusCode.BadRequest;
-                await response.WriteAsJsonAsync(new CreationResponseDTO { Success = false, Message = "Invalid payload" });
+                _logger.LogError("Something unexpected happenned!");
```
We know the cause from the exception, so we should concat the `e.Message` to the logger's message.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs.orig:
```diff
@@ -0,0 +1,57 @@
+using Microsoft.Extensions.Logging;
+
+namespace AzureEndPointReaction.Functions.Questionnaires
+{
```
Something went wrong with resolving the conflicts, this has no place here.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireDeletionWorkerEncapsulator.cs:
```diff
@@ -31,15 +31,53 @@ public sealed class QuestionnaireDeletionWorkerEncapsulator(IQuestionnaireServic
+        [OpenApiResponseWithoutBody(HttpStatusCode.InternalServerError)]
+        public async Task<HttpResponseData> ExecuteTaskAsync(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "delete", Route = "questionnaires/{id:guid}")] HttpRequestData request,
+            FunctionContext context)
```
Instead of FunctionContext, you can use Guid id, and it gets binded automatically, so you don't need this whole ceremony getting the id from the context.

SzLRichard replied:
I wanted to do that, but the functions api endpoints implement the IQuestionnaireWorker interface. This is why I was unable to add `Guid id` as a parameter, and instead of modifying that I found a workaround.

BHunor-ex replied:
Is the IQuestionnaireWorker is an interface outside of our control (I mean didn't you create it)?

If the answer is no (which i think it is), the FunctionContext should be removed from there as well.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs:
```diff
@@ -31,22 +33,33 @@ public sealed class QuestionnaireCompilerWorkerEncapsulator(IQuestionnaireServic
             bodyType: typeof(CreationResponseDTO) // replace dto
             )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "questionnaires")] HttpRequestData request, FunctionContext context, CancellationToken token)
+        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "questionnaires")] HttpRequestData request, FunctionContext context)
```
FunctionContext is not needed as I see, it doesn't need to appear i guess.

BHunor-ex commented on FeedBackApp.Backend/Application/Services/Interfaces/IQuestionnaireWorker.cs:
```diff
@@ -5,6 +5,6 @@ namespace Application.Services.Interfaces
     public interface IQuestionnaireWorker
     {
-        Task<HttpResponseData> ExecuteTaskAsync(HttpRequestData request, FunctionContext context, CancellationToken token);
+        Task<HttpResponseData> ExecuteTaskAsync(HttpRequestData request, FunctionContext context);
```
I don't think FunctionContext is needed.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireEvaluationWorkerEncapsulator.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class QuestionnaireEvaluationWorkerEncapsulator(IEvaluationService
             bodyType: typeof(object) // replace with actual evaluation DTO
         )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "evaluations/{id:guid}")] HttpRequestData request, FunctionContext context, CancellationToken token)
+        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "evaluations/{id:guid}")] HttpRequestData request, FunctionContext context)
```
remove FunctionContext, add Guid id.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireSummaryRequestWorkerEncapsulator.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class QuestionnaireSummaryRequestWorkerEncapsulator(IQuestionnaire
             bodyType: typeof(object) // replace with `summary` DTO
         )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "questionnaires/{id:guid}")] HttpRequestData request, FunctionContext context, CancellationToken token)
+        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "questionnaires/{id:guid}")] HttpRequestData request, FunctionContext context)
```
remove FunctionContext, add Guid id.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireUpdateRequestWorkerEncapsulator.cs:
```diff
@@ -32,14 +32,19 @@ public sealed class QuestionnaireUpdateRequestWorkerEncapsulator(IEvaluationServ
+            contentType: "application/json",
+            bodyType: typeof(UpdateResponseDTO)
+        )]
+        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "patch", Route = "evaluations/{id:guid}")] HttpRequestData request, FunctionContext context)
```
remove FunctionContext, add Guid id.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/Program.cs.orig:
```diff
@@ -0,0 +1,91 @@
+﻿using Application.Services;
```
This should be deleted

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Backend.Infrastructure/Persistence/Repository/QuestionnaireRepository.cs:
```diff
@@ -13,26 +14,96 @@ public QuestionnaireRepository(AppDBContext context)
+
+            _context.Add(metadata);
+            _context.Add(tempForSave);
+            await _context.SaveChangesAsync();
```
We should call .SaveChangesAsync() only once, it's a costly operation.

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Core/Model/QuestionAnswer.cs:
```diff
@@ -1,15 +1,10 @@
-        [JsonPropertyName("answer")]
+        public string QuestionId { get; set; } = string.Empty;
+
         public string? Answer { get; set; } = null;
```
Can the answer be null? Shouldn't it be emtpy in those cases?

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Core/Model/QuestionTemplate.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public string Id { get; set; } = string.Empty;
+        public string Question { get; set; } = string.Empty;
+        public QuestionType? Type { get; set; }
```
Same as in the DTO, QuestionType shouldn't be null, it's safer to add a  QuestionType.Unknown / QuestionType.NoType (or something like that)

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Core/Model/QuestionTemplate.cs:
```diff
@@ -0,0 +1,13 @@
+        public string Id { get; set; } = string.Empty;
+        public string Question { get; set; } = string.Empty;
+        public QuestionType? Type { get; set; }
+        public IList<string>? AnswerOptions { get; set; } = new List<string>();
```
Same as in the DTO,  AnswerOptions shouldn't be null, but an EmptyList maybe.

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Core/Repositories/IQuestionnaireRepository.cs:
```diff
@@ -5,9 +5,9 @@ namespace FeedBackApp.Core.Repositories
-        Task SaveQuestionnaireAsync(Questionnaire questionnaire);
-        Task<Metadata?> GetMetadataAsync(string id);
-        Task<Questionnaire?> GetQuestionnaireAsync(string id);
+        public Task CompileAndSaveAsync(SurveyMetadata metadata);
```
public is not needed here.

SzLRichard committed:
Added requested changes

@BHunor-ex

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs:
```diff
@@ -11,7 +11,7 @@
 
 namespace AzureEndPointReaction.Functions.Questionnaires
 {
-    public sealed class QuestionnaireCompilerWorkerEncapsulator(IQuestionnaireService service, ILogger<QuestionnaireCompilerWorkerEncapsulator> logger, IEmailService emailService) : IQuestionnaireWorker
```
I didn't meant to remove the Interface, but to remove the FunctionContext from the interface, and put the Guid id there.

SzLRichard replied:
This is the only endpoint right now that doesn't need a `Guid id`.  Should I still implement an interface that has a `Guid id` parameter?

BHunor-ex replied:
Will it need an id in the future? If not  then no.

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireCompilerWorkerEncapsulator.cs:
```diff
@@ -33,14 +33,15 @@ public sealed class QuestionnaireCompilerWorkerEncapsulator(IQuestionnaireServic
             contentType: "application/json", 
             bodyType: typeof(CreationResponseDTO) // replace dto
             )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "questionnaires")] HttpRequestData request, FunctionContext context)
```
same here

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireDeletionWorkerEncapsulator.cs:
```diff
@@ -10,7 +10,7 @@
 
 namespace AzureEndPointReaction.Functions.Questionnaires
 {
-    public sealed class QuestionnaireDeletionWorkerEncapsulator(IQuestionnaireService service, ILogger<QuestionnaireDeletionWorkerEncapsulator> logger) : IQuestionnaireWorker
```
here aswell

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireEvaluationWorkerEncapsulator.cs:
```diff
@@ -10,7 +10,7 @@
 
 namespace AzureEndPointReaction.Functions.Questionnaires
 {
-    public sealed class QuestionnaireEvaluationWorkerEncapsulator(IEvaluationService service, ILogger<QuestionnaireEvaluationWorkerEncapsulator> logger) : IQuestionnaireWorker
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireEvaluationWorkerEncapsulator.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class QuestionnaireEvaluationWorkerEncapsulator(IEvaluationService
             contentType: "application/json",
             bodyType: typeof(object) // replace with actual evaluation DTO
         )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "evaluations/{id:guid}")] HttpRequestData request, FunctionContext context)
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireSummaryRequestWorkerEncapsulator.cs:
```diff
@@ -10,7 +10,7 @@
 
 namespace AzureEndPointReaction.Functions.Questionnaires
 {
-    public sealed class QuestionnaireSummaryRequestWorkerEncapsulator(IQuestionnaireService service, ILogger<QuestionnaireSummaryRequestWorkerEncapsulator> logger) : IQuestionnaireWorker
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireSummaryRequestWorkerEncapsulator.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class QuestionnaireSummaryRequestWorkerEncapsulator(IQuestionnaire
             contentType: "application/json",
             bodyType: typeof(object) // replace with `summary` DTO
         )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "questionnaires/{id:guid}")] HttpRequestData request, FunctionContext context)
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireUpdateRequestWorkerEncapsulator.cs:
```diff
@@ -10,7 +10,7 @@
 
 namespace AzureEndPointReaction.Functions.Questionnaires
 {
-    public sealed class QuestionnaireUpdateRequestWorkerEncapsulator(IEvaluationService service, ILogger<QuestionnaireUpdateRequestWorkerEncapsulator> logger) : IQuestionnaireWorker
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/AzureEndPointReaction/Functions/Questionnaires/QuestionnaireUpdateRequestWorkerEncapsulator.cs:
```diff
@@ -29,15 +29,15 @@ public sealed class QuestionnaireUpdateRequestWorkerEncapsulator(IEvaluationServ
             contentType: "application/json",
             bodyType: typeof(UpdateResponseDTO)
         )]
-        public async Task<HttpResponseData> ExecuteTaskAsync([HttpTrigger(AuthorizationLevel.Anonymous, "patch", Route = "evaluations/{id:guid}")] HttpRequestData request, FunctionContext context)
```
same

BHunor-ex commented on FeedBackApp.Backend/AzureFunctionsAPI/Program.cs:
```diff
@@ -57,11 +57,11 @@
         services.AddScoped<IEvaluationService, EvaluationService>();
         services.AddScoped<IQuestionnaireRepository, QuestionnaireRepository>();
         services.AddScoped<IQuestionnaireService, QuestionnaireService>();
-        services.AddScoped<IQuestionnaireWorker, QuestionnaireCompilerWorkerEncapsulator>();
```
same

BHunor-ex commented on FeedBackApp.Backend/FeedBackApp.Backend.Infrastructure/Persistence/Repository/QuestionnaireRepository.cs:
```diff
@@ -23,7 +23,6 @@ public async Task CompileAndSaveAsync(SurveyMetadata metadata)
 
             _context.Add(metadata);
             _context.Add(tempForSave);
-            await _context.SaveChangesAsync();
```
I can't  add comment to the 48th line, but since the Answer cannot be null from now, we should set it empty  no?

BHunor-ex reviewed (changes requested):
There are some github action warnings, please look at them.

SzLRichard committed:
Brought back IQuestionnaireWorker

Brought back the IQuestionnaireWorker interface for each function, except for the one that doesn't need a Guid. Also fixed workflow warnings.
